spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/ReactiveDatastoreTemplate.java

```diff
@@ -0,0 +1,205 @@
+		}
+
+		return convertAndResolveRelationshipsEagerly(
+				Flux.fromIterable(
+						() -> this.datastoreTemplate.getDatastoreReadWriter().get(keys.toArray(new Key[] {}))),
+				entityClass, context);
+	}
+
+	public <T> Flux<T> findAll(Class<T> entityClass, DatastoreQueryOptions queryOptions) {
+		DatastorePersistentEntity<?> persistentEntity = this.datastoreTemplate.getDatastoreMappingContext()
+				.getPersistentEntity(entityClass);
+		EntityQuery.Builder builder = Query.newEntityQueryBuilder()
+				.setKind(persistentEntity.kindName());
+		DatastoreTemplate.applyQueryOptions(builder, queryOptions, persistentEntity);
+		Query query = builder.build();
+
+		return convertAndResolveRelationshipsEagerly(
+				Flux.fromIterable(() -> this.datastoreTemplate.getDatastoreReadWriter().run(query)), entityClass,
+				new DatastoreTemplate.ReadContext());
+	}
+
+	public <T> Flux<T> query(Query<? extends BaseEntity> query, Class<T> entityClass) {
+		return convertAndResolveRelationshipsEagerly(
+				(Flux<Entity>) Flux.fromIterable(() -> this.datastoreTemplate.getDatastoreReadWriter().run(query)),
+				entityClass, new DatastoreTemplate.ReadContext());
+	}
+
+	public <A, T> Flux<T> query(Query<A> query, Function<A, T> entityFunc) {
+		return Flux.fromIterable(() -> this.datastoreTemplate.getDatastoreReadWriter().run(query))
+				.map(entityFunc);
+	}
+
+	public Flux<Key> queryKeys(Query<Key> query) {
+		return Flux.fromIterable(() -> this.datastoreTemplate.getDatastoreReadWriter().run(query));
+	}
+
+	public <T> Flux<T> findAll(Class<T> entityClass) {
+		return findAll(entityClass, null);
+	}
+
+	public <T> Flux<T> queryByExample(Example<T> example, DatastoreQueryOptions queryOptions) {
+		return query(this.datastoreTemplate.exampleToQuery(example, queryOptions, false), example.getProbeType());
+	}
+
+	public <T> Flux<Key> keyQueryByExample(Example<T> example, DatastoreQueryOptions queryOptions) {
+		Query query = this.datastoreTemplate.exampleToQuery(example, queryOptions, true);
+		return Flux.fromIterable(() -> this.datastoreTemplate.getDatastoreReadWriter().run(query));
+	}
+
+	public <T> Mono<Boolean> existsById(Object id, Class<T> entityClass) {
+		return findById(id, entityClass).map(Objects::nonNull);
+	}
+
+	private Flux<Key> findAllKeys(Class entityClass) {
+		return queryKeys(Query.newKeyQueryBuilder().setKind(
+				this.datastoreTemplate.getDatastoreMappingContext()
+						.getPersistentEntity(entityClass).kindName())
+				.build());
+	}
+
+	private <A> Flux<A> convertAndResolveRelationshipsEagerly(Flux<Entity> entities, Class<A> entityClass,
+			DatastoreTemplate.ReadContext readContext) {
+		DatastorePersistentEntity datastorePersistentEntity = this.datastoreTemplate.getDatastoreMappingContext()
+				.getPersistentEntity(entityClass);
+		return entities.doOnNext(entity -> readContext.putReadEntity(entity.getKey(), entity))
+				.map(entity -> this.datastoreTemplate.convertEntityResolveDescendantsAndReferences(entityClass,
```

Comment:
look at pubsub example, get another thread to do this single entity conversion. "executorservice". 
some sort of "zip" that combines parent and children when both are ready.